### PR TITLE
NodeMapV2: Introduce a V2 of the NodeMap which includes SPI values 

### PR DESCRIPTION
**Pull request by ldelossa**
This pull request adds a new V2 addition to the NodeMap. 

The NodeMap is responsible for inventorying and restoration of allocated NodeIDs which are subsequently used to correctly match XFRM policies and states given the ultimate destination of IPSec encrypted traffic.  

The v2 edition of NodeMap now associates each node entry with their SPI which is also used to correctly match XFRM policies and states given the ultimate destination of IPSec encrypted traffic. 

However, by providing the SPI in NodeMap we can now query the SPI value associated with a Cilium managed node. Prior to this we could only query SPI on a per-pod basis.

This ability becomes especially useful when combined with the [Encrypted Overlay feature](https://github.com/cilium/cilium/pull/31073), and provides us a way to retrieve SPI when only the destination Node address is obtainable, as opposed to the destination pod's address. 

This PR includes an upgrade migration which will copy v1 map entries into the v2 map and fill in the current SPI.
This migration works on the assumption that the cluster has a stable SPI and [no pending key rotation is occurring](https://github.com/cilium/cilium/pull/31437) during the migration (Cilium upgrade in other words). 

It is best to review this PR one commit at a time. 

```release-note
Transition to NodeMapV2 which now includes SPI in its map values.
```

### CODE REVIEW

**Review by dylandreimerink — APPROVED** [review]

**Review by thorn3r — APPROVED** [review]
LGTM! In the case of a downgrade, a new nodemap v1 would be created with new entries, right?

File: bpf/lib/maps.h
Comment (pchaigno) [inline bpf/lib/maps.h:198]: I jumped straight here to ask for padding and it's already here :heart:

(We probably will need this padding at some point to have a flag that indicates the type of node IP, to help with masquerading. Best to migrate once now :sweat_smile: )

File: pkg/datapath/linux/node_ids.go
Comment (pchaigno) [inline pkg/datapath/linux/node_ids.go:88]: Hm, this defines whether we want to force an update of the node ID map because we know the SPI changed. Could it happen that we don't have an oldNode but the node ID map still contains an entry for that node IP? If so, we might want to default to true.

Maybe one such scenario is IP reuse after agent is down:
1. NodeA with IP_Z is created, mapping created.
2. Agent goes down.
3. NodeA is deleted.
4. Agent comes back.
5. SPI is incremented via a key rotation.
6. NodeB with IP_Z is created.

At 6 we want to update the existing entry in the node ID map, no?

File: pkg/maps/nodemap/node_map_v2.go
Comment (youngnick) [inline pkg/maps/nodemap/node_map_v2.go]: Feels like we should be avoiding a panic just in case the type assertion fails?

```suggestion
			key, ok := k.(*NodeKey)
			if !ok {
				// Some key error handling here
			}
			value, ok := v.(*NodeValueV2)
			if !ok {
				// Some value error handling here
			}
```

**Review by youngnick — COMMENTED** [review]
LGTM aside from one question about type assertions.

File: pkg/maps/nodemap/node_map_v2.go
Comment (ldelossa) [reply pkg/maps/nodemap/node_map_v2.go]: Think this is totally logical, will update.

File: pkg/logging/logfields/logfields.go
Comment (pchaigno) [inline pkg/logging/logfields/logfields.go]: We already have SPI as a field here. I'd rather stick to one consistent name for this in logs.

File: pkg/maps/nodemap/node_map_v2.go
Comment (pchaigno) [inline pkg/maps/nodemap/node_map_v2.go:30]: Call me crazy, but if we named the existing Go types for the node ID map `Old` instead of naming the new types `V2`, wouldn't that significantly reduce how many places you need to touch all over the codebase? It would also mean we wouldn't be dragging a `V2` around for the next 10 releases :sweat_smile:

File: pkg/maps/nodemap/node_map_v2.go
Comment (pchaigno) [inline pkg/maps/nodemap/node_map_v2.go:24]: I think we also need to ensure that we dump both maps in the bugtool. Just run `git grep -W v2 bugtool/` and you'll see what I mean :sweat_smile:

**Review by pchaigno — COMMENTED** [review]
This looks pretty good to me overall. A couple concerns below I want to discuss so not approving just yet :grin: 

You know me, I would have also liked even smaller commits for the commit history :innocent: 

I think regardless of reviews, we'll want to rebase to catch https://github.com/cilium/cilium/pull/31637.

File: pkg/maps/nodemap/node_map_v2.go
Comment (youngnick) [reply pkg/maps/nodemap/node_map_v2.go:30]: Oh, this is a good idea, +1!

Comment (ldelossa) [reply pkg/maps/nodemap/node_map_v2.go:30]: My opinion here is to be very explicit about a new version of a map being introduced. 

In my mind, it makes things very explicit that the map has transitioned and can help with some meta-debugging, knowing that map is new. 

I don't quite see whats wrong with keeping the V2 naming around, indefinitely? There is no reason to "hide" that this map went through more then one revision? 

I also did just find it a bit saner to introduce the new name, find all references, and evaluate each reference explicitly. Since this forced me to look at every usage of the old map and think about what will occur when the new map is used.

Comment (pchaigno) [reply pkg/maps/nodemap/node_map_v2.go:30]: The issue isn't so much `V2` staying around; that's just the cherry on the cake if we can get rid of it. The issue IMO is the potential for backport hell if we ever need to backport this anywhere.

I understand the desire to be explicit about the new version. I'm not suggesting changing the map name on the filesystem; that can remain `v2` if we prefer. I'm only talking about the Golang side of things where the `V2` suffix change is a little bit "invasive".

**Comment by ldelossa** [comment]
@thorn3r 

> LGTM! In the case of a downgrade, a new nodemap v1 would be created with new entries, right?

A downgrade PR will be opened again v1.15 stable to address this situation after this is merged.

File: pkg/maps/nodemap/node_map_v2.go
Comment (ldelossa) [reply pkg/maps/nodemap/node_map_v2.go:30]: I'm not sure how much the approach reduces the diff created, since we need to update all the areas where MapV1 is used since the fundamental way we use the mapValue now changes. Prior to this, we just used uint16 almost everywhere as the map Value. 

I can give it a try tho, and see if it produces much less of a diff while not making the code confusing.

File: pkg/logging/logfields/logfields.go
Comment (ldelossa) [reply pkg/logging/logfields/logfields.go]: Thanks, this was just an oversight. Will fix.

File: pkg/datapath/linux/node_ids.go
Comment (ldelossa) [reply pkg/datapath/linux/node_ids.go:88]: Yeah this is a really good point. The code I have here is using IP like a stable ID. 

I agree, changing it to default to true is probably simplest. We may get some spurious force updates to the map, but those should be minimal.

Comment (ldelossa) [reply pkg/datapath/linux/node_ids.go:88]: This is kinda making me think of a bigger point. 

Do we cleanup eBPF Node map at all? For instance, what occurs in the same scenario above except NodeB uses a completely separate IP. Does NodeA's IP stay in the map indefinitely until something reuses it?

File: pkg/maps/nodemap/node_map_v2.go
Comment (ldelossa) [reply pkg/maps/nodemap/node_map_v2.go]: Resolved in latest push

File: pkg/datapath/linux/node_ids.go
Comment (ldelossa) [reply pkg/datapath/linux/node_ids.go:88]: Original issue fixed in latest push.

File: pkg/logging/logfields/logfields.go
Comment (ldelossa) [reply pkg/logging/logfields/logfields.go]: Fixed in latest push

File: bpf/lib/maps.h
Comment (asauber) [reply bpf/lib/maps.h:198]: Based on a quick look at the ESP spec, SPI can be up to 32 bits. 

If at some point we want pod-unique encryption parameters, it might make sense to bump SPI up to 16 or 32 bits. I'm not familiar with how that would affect the rest of the IPSec datapath.

Given the general purpose nature of SPI, we could also consider this 8 bit padding that type-specific component of the SPI, not needing any changes.

File: cilium-dbg/cmd/bpf_nodeid_list.go
Comment (asauber) [inline cilium-dbg/cmd/bpf_nodeid_list.go:52]: I'll echo @pchaigno that this command should attempt to dump both maps until the old one is no longer supported.

**Review by asauber — APPROVED** [review]
Overall looks good to me. A few thoughts.

1. I'm glad that we addressed a possible bug due to IP reuse. Let's keep that type of issue in mind.
2. I think that it is helpful to name the new map and types with a V2 suffix, even for the sake of rebasing, because it gets eyes on those areas of the codebase (it can be a slog but it's also often worth it). Also "explicit is better than implicit".
3. Looking forward to approving the PR for the downgrade path :)

File: pkg/maps/nodemap/node_map_v2.go
Comment (ldelossa) [reply pkg/maps/nodemap/node_map_v2.go:30]: @pchaigno do you feel strongly enough about this to block the PR. 

I was working thru this change and I ran into a few things that don't sit well with me.
1. Going back and renaming the existing references feels a bit odd, instead of leaving the existing implementation and introducing v2. 
2. When we begin to try to rename the old map and use the existing map name, we destroy the commit structure I was after, where one commit cleanly introduces the new map, and another does the cut over. This needs to be smashed into one large commit if we want the commit to build, not quite a fan of this. 
3. I strongly believe any issues of repetitive rebasing can be fixed by enable `git rerere`. This solves most of the annoying rebase issues I run into when backporting.

**Review by jschwinger233 — COMMENTED** [review]
Maybe we want to exclude `bpf/testdata/unreachable-tailcall.o`

File: pkg/maps/nodemap/node_map_v2.go
Comment (youngnick) [reply pkg/maps/nodemap/node_map_v2.go:30]: To be clear, "it sounds like a good idea" is not blocking feedback from me, so it's up to @pchaigno.

**Comment by ldelossa** [comment]
@jschwinger233 

We are in the process of making that object file produce a diff less often. See: https://github.com/cilium/cilium/pull/31644

The explanation for why we can't just put this object file in .gitignore is because when the object file is not there, it breaks a plain run of `go test` and this would be annoying for most.

**Comment by ldelossa** [comment]
Alright, this PR is about to take a U-Turn. 

After internal discussion, we need this change present in v1.15. 

Therefore, we need to slam a map migration in the middle of a patch release. 

To accomplish this without any major headaches, this PR will change to keep MapV1 around and 'mirroring' any writes to MapV2 into MapV1. 

That way, if a user then downgrades, MapV1 is still present and up to date with any changes that occurred while MapV2 was present. 

Stay tuned.

File: pkg/maps/nodemap/node_map_v2.go
Comment (ldelossa) [reply pkg/maps/nodemap/node_map_v2.go:24]: @pchaigno it appears we don't dump the node maps at all currently, looking at the file that the grep you suggested. I don't think the nodemap is dumped?

**Comment by ldelossa** [comment]
/test-backport-1.15

**Comment by harsimran-pabla** [comment]
Hi @ldelossa 👋  Is this binary file pushed by mistake ? 
[pkg/bpf/testdata/unreachable-tailcall.o](https://github.com/cilium/cilium/pull/31431/commits/4addd422af53634aac0ff6e4a2caf20552e9dd7d#diff-e286267cdbb00d2ef467e9f7a2b83ac3f309e0fd45c01d1211ca25c2bf29328e)

**Comment by ldelossa** [comment]
@harsimran-pabla Nope, see: 
https://github.com/cilium/cilium/pull/31431#issuecomment-2024922548

**Comment by harsimran-pabla** [comment]
> @harsimran-pabla Nope, see: [#31431 (comment)](https://github.com/cilium/cilium/pull/31431#issuecomment-2024922548)

I should have read the comments :)

File: bpf/tests/ipsec_from_lxc_generic.h
Comment (asauber) [inline bpf/tests/ipsec_from_lxc_generic.h:133]: Do these map updates need to be mirrored as well?

Comment (ldelossa) [reply bpf/tests/ipsec_from_lxc_generic.h:133]: We don't update the NodeMap during cilium runtime from the datapath, so its okay to not have any mirroring here. The datapath won't care much about MapV1's existence.

File: pkg/maps/nodemap/node_map_v2.go
Comment (pchaigno) [reply pkg/maps/nodemap/node_map_v2.go:24]: Oh, crap, you're right! Let's handle that in an ad-hoc PR then.

File: pkg/maps/nodemap/node_map_v2.go
Comment (pchaigno) [reply pkg/maps/nodemap/node_map_v2.go:30]: We discussed this offline and concluded that conflicts are unlikely outside of the normal backporting path to v1.15, which Louis will likely handle.

File: pkg/maps/nodemap/node_map_v2.go
Comment (pchaigno) [inline pkg/maps/nodemap/node_map_v2.go]: Isn't that going to get very noisy on large clusters? Maybe enough to log how many entries were migrated and the SPI?

**Review by pchaigno — APPROVED** [review]

File: cilium-dbg/cmd/bpf_nodeid_list.go
Comment (pchaigno) [reply cilium-dbg/cmd/bpf_nodeid_list.go:52]: I don't think I said that :sweat_smile: 

I've opened a PR to dump the two raw maps (via bpftool), but I don't think it would bring much to have this command dump and pretty print entries from both maps. It could even be confusing during investigations.

**Comment by pchaigno** [comment]
@ldelossa I've labeled the PR according to our Slack discussion.

**Comment by ldelossa** [comment]
@pchaigno 

After some further conversation with community team, we wont be back porting this into v1.15.

File: pkg/maps/nodemap/node_map_v2.go
Comment (ldelossa) [reply pkg/maps/nodemap/node_map_v2.go]: Do we mind the noise when it only occurs once at startup?

Comment (ldelossa) [reply pkg/maps/nodemap/node_map_v2.go]: Not opposed to reducing it tho, wonder if seeing each specific value is worth it. 

Its true tho, itll blow up the output when theres 40+ nodes. 

Okay i think ill reduce it.

Comment (ldelossa) [reply pkg/maps/nodemap/node_map_v2.go]: Resolved in latest push.

File: cilium-dbg/cmd/bpf_nodeid_list.go
Comment (ldelossa) [reply cilium-dbg/cmd/bpf_nodeid_list.go:52]: Yeah, we don't want the CLI to dump both maps. Rather, we want bugtool to dump the bytes of both maps during systemdump.

However, this isn't occurring as of today see: https://github.com/cilium/cilium/pull/31431#discussion_r1542124198

I'll open a PR which dumps both maps in our systemdumps after this PR.

Comment (ldelossa) [reply cilium-dbg/cmd/bpf_nodeid_list.go:52]: Oh, sorry, I think Paul opened this PR already  :)

File: pkg/maps/nodemap/node_map_v2.go
Comment (pchaigno) [reply pkg/maps/nodemap/node_map_v2.go]: Thanks! We may occasionally enable debug mode in clusters with hundreds or even thousands of nodes so that's why I thought it was probably worth reducing.

File: bpf/lib/maps.h
Comment (julianwiedmann) [inline bpf/lib/maps.h:199]: Could you please also add this to the alignchecker? :pray:

Comment (julianwiedmann) [reply bpf/lib/maps.h:199]: https://github.com/cilium/cilium/pull/35309

**Comment by julianwiedmann** [comment]
> @thorn3r
> 
> > LGTM! In the case of a downgrade, a new nodemap v1 would be created with new entries, right?
> 
> A downgrade PR will be opened again v1.15 stable to address this situation after this is merged.

@ldelossa did that v1.15 downgrade toleration happen after all, or is it still todo?

**Comment by julianwiedmann** [comment]
> > @thorn3r
> > > LGTM! In the case of a downgrade, a new nodemap v1 would be created with new entries, right?
> > 
> > 
> > A downgrade PR will be opened again v1.15 stable to address this situation after this is merged.
> 
> @ldelossa did that v1.15 downgrade toleration happen after all, or is it still todo?

Ah, I think that's now fully handled by the "mirror write" in `nodeMapV2`'s `Update()` routine, which ensures the v1 map is kept in-sync and can be used after a downgrade to `v1.15`. So right now a downgrade would just "leak" the unused v2 map?

**Pull request merged**
